### ISSUE TITLE
certbot-auto: LookupError: unknown encoding: idna

### ISSUE THREAD

**Issue by fdellwing** [certbot/certbot#5358]
## My operating system is (include version):

Distributor ID: Ubuntu
Description:    Ubuntu 12.04.5 LTS
Release:        12.04
Codename:       precise
Kernel: 3.2.0

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):

certbot-auto

## I ran this command and it produced this output:

```
[11:06 root@drupal-cms ~] > ./certbot-auto --apache -d www.bagusche.com
Bootstrapping dependencies for Debian-based OSes... (you can skip this with --no-bootstrap)
OK   http://de.archive.ubuntu.com precise Release.gpg
OK   http://de.archive.ubuntu.com precise-updates Release.gpg              
OK   http://de.archive.ubuntu.com precise-backports Release.gpg            
OK   http://security.ubuntu.com precise-security Release.gpg               
OK   http://ppa.launchpad.net precise Release.gpg  
OK   http://de.archive.ubuntu.com precise Release                                                    
OK   http://de.archive.ubuntu.com precise-updates Release                                            
OK   http://security.ubuntu.com precise-security Release                                             
OK   http://de.archive.ubuntu.com precise-backports Release                                          
OK   http://ppa.launchpad.net precise Release.gpg                         
OK   http://de.archive.ubuntu.com precise/main i386 Packages                                        
OK   http://de.archive.ubuntu.com precise/restricted i386 Packages                                                          
OK   http://de.archive.ubuntu.com precise/universe i386 Packages                                                            
OK   http://de.archive.ubuntu.com precise/multiverse i386 Packages                                                          
OK   http://de.archive.ubuntu.com precise/main TranslationIndex                                                             
OK   http://de.archive.ubuntu.com precise/multiverse TranslationIndex                                                       
OK   http://de.archive.ubuntu.com precise/restricted TranslationIndex                                                       
OK   http://de.archive.ubuntu.com precise/universe TranslationIndex                                                         
OK   http://security.ubuntu.com precise-security/main i386 Packages                                                         
OK   http://de.archive.ubuntu.com precise-updates/main i386 Packages   
OK   http://de.archive.ubuntu.com precise-updates/restricted i386 Packages                     
OK   http://de.archive.ubuntu.com precise-updates/universe i386 Packages                       
OK   http://de.archive.ubuntu.com precise-updates/multiverse i386 Packages                     
OK   http://de.archive.ubuntu.com precise-updates/main TranslationIndex                        
OK   http://de.archive.ubuntu.com precise-updates/multiverse TranslationIndex                  
OK   http://security.ubuntu.com precise-security/restricted i386 Packages                      
OK   http://security.ubuntu.com precise-security/universe i386 Packages                        
OK   http://security.ubuntu.com precise-security/multiverse i386 Packages
OK   http://security.ubuntu.com precise-security/main TranslationIndex 
OK   http://security.ubuntu.com precise-security/multiverse TranslationIndex
OK   http://security.ubuntu.com precise-security/restricted TranslationIndex
OK   http://security.ubuntu.com precise-security/universe TranslationIndex
OK   http://de.archive.ubuntu.com precise-updates/restricted TranslationIndex
OK   http://de.archive.ubuntu.com precise-updates/universe TranslationIndex
OK   http://de.archive.ubuntu.com precise-backports/main i386 Packages 
OK   http://de.archive.ubuntu.com precise-backports/restricted i386 Packages
OK   http://de.archive.ubuntu.com precise-backports/universe i386 Packages                     
OK   http://de.archive.ubuntu.com precise-backports/multiverse i386 Packages                   
OK   http://de.archive.ubuntu.com precise-backports/main TranslationIndex                      
OK   http://de.archive.ubuntu.com precise-backports/multiverse TranslationIndex                
OK   http://de.archive.ubuntu.com precise-backports/restricted TranslationIndex                
OK   http://security.ubuntu.com precise-security/main Translation-en                           
OK   http://security.ubuntu.com precise-security/multiverse Translation-en                     
OK   http://security.ubuntu.com precise-security/restricted Translation-en                     
OK   http://de.archive.ubuntu.com precise-backports/universe TranslationIndex                  
OK   http://de.archive.ubuntu.com precise/main Translation-de                                  
OK   http://de.archive.ubuntu.com precise/main Translation-en                                  
OK   http://de.archive.ubuntu.com precise/multiverse Translation-de                            
OK   http://de.archive.ubuntu.com precise/multiverse Translation-en                            
OK   http://de.archive.ubuntu.com precise/restricted Translation-de                            
OK   http://security.ubuntu.com precise-security/universe Translation-en                       
OK   http://ppa.launchpad.net precise Release.gpg                                              
OK   http://de.archive.ubuntu.com precise/restricted Translation-en                              
OK   http://de.archive.ubuntu.com precise/universe Translation-de                                
OK   http://de.archive.ubuntu.com precise/universe Translation-en                                
OK   http://de.archive.ubuntu.com precise-updates/main Translation-en                            
OK   http://de.archive.ubuntu.com precise-updates/multiverse Translation-en                      
OK   http://de.archive.ubuntu.com precise-updates/restricted Translation-en                      
OK   http://de.archive.ubuntu.com precise-updates/universe Translation-en
OK   http://de.archive.ubuntu.com precise-backports/main Translation-en  
OK   http://de.archive.ubuntu.com precise-backports/multiverse Translation-en
OK   http://de.archive.ubuntu.com precise-backports/restricted Translation-en
OK   http://ppa.launchpad.net precise Release                            
OK   http://de.archive.ubuntu.com precise-backports/universe Translation-en                                             
OK   https://apt.dockerproject.org ubuntu-precise Release.gpg                                                           
OK   https://apt.dockerproject.org ubuntu-precise Release                
OK   https://apt.dockerproject.org ubuntu-precise/main i386 Packages  
Hole:1 https://apt.dockerproject.org ubuntu-precise/main TranslationIndex [454 B]
Ign https://apt.dockerproject.org ubuntu-precise/main TranslationIndex          
OK   http://ppa.launchpad.net precise Release                                   
OK   http://ppa.launchpad.net precise Release                                                   
OK   http://ppa.launchpad.net precise/main Sources                                              
OK   http://ppa.launchpad.net precise/main i386 Packages                 
OK   http://ppa.launchpad.net precise/main TranslationIndex              
OK   http://ppa.launchpad.net precise/main Sources                       
OK   http://ppa.launchpad.net precise/main i386 Packages                 
OK   http://ppa.launchpad.net precise/main TranslationIndex              
OK   http://ppa.launchpad.net precise/main Sources                       
OK   http://ppa.launchpad.net precise/main i386 Packages                 
OK   http://ppa.launchpad.net precise/main TranslationIndex              
OK   http://ppa.launchpad.net precise/main Translation-en
OK   http://ppa.launchpad.net precise/main Translation-en
OK   http://ppa.launchpad.net precise/main Translation-en
Ign https://apt.dockerproject.org ubuntu-precise/main Translation-de_DE
Ign https://apt.dockerproject.org ubuntu-precise/main Translation-de
Ign https://apt.dockerproject.org ubuntu-precise/main Translation-en
Paketlisten werden gelesen... Fertig
Paketlisten werden gelesen... Fertig
Abhängigkeitsbaum wird aufgebaut       
Statusinformationen werden eingelesen... Fertig
gcc ist schon die neueste Version.
libffi-dev ist schon die neueste Version.
python-virtualenv ist schon die neueste Version.
ca-certificates ist schon die neueste Version.
python ist schon die neueste Version.
python-dev ist schon die neueste Version.
augeas-lenses ist schon die neueste Version.
libaugeas0 ist schon die neueste Version.
openssl ist schon die neueste Version.
libssl-dev ist schon die neueste Version.
0 aktualisiert, 0 neu installiert, 0 zu entfernen und 0 nicht aktualisiert.
Creating virtual environment...
Installing Python packages...
Had a problem while installing Python packages.

pip prints the following errors: 
=====================================================
Collecting argparse==1.4.0 (from -r /tmp/tmp.gF2NdgWFHM/letsencrypt-auto-requirements.txt (line 11))
Exception:
Traceback (most recent call last):
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/commands/install.py", line 324, in run
    requirement_set.prepare_files(finder)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/req/req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/req/req_set.py", line 554, in _prepare_file
    require_hashes
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/req/req_install.py", line 278, in populate_link
    self.link = finder.find_requirement(self, upgrade)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/index.py", line 465, in find_requirement
    all_candidates = self.find_all_candidates(req.name)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/index.py", line 423, in find_all_candidates
    for page in self._get_pages(url_locations, project_name):
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/index.py", line 568, in _get_pages
    page = self._get_page(location)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/index.py", line 683, in _get_page
    return HTMLPage.get_page(link, session=self.session)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/index.py", line 792, in get_page
    "Cache-Control": "max-age=600",
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/_vendor/requests/sessions.py", line 488, in get
    return self.request('GET', url, **kwargs)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/download.py", line 386, in request
    return super(PipSession, self).request(method, url, *args, **kwargs)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/_vendor/requests/sessions.py", line 461, in request
    prep = self.prepare_request(req)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/_vendor/requests/sessions.py", line 394, in prepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/_vendor/requests/models.py", line 294, in prepare
    self.prepare_url(url, params)
  File "/opt/eff.org/certbot/venv/local/lib/python2.7/site-packages/pip/_vendor/requests/models.py", line 361, in prepare_url
    host = host.encode('idna').decode('utf-8')
LookupError: unknown encoding: idna
=====================================================

Certbot has problem setting up the virtual environment.

We were not be able to guess the right solution from your pip 
output.

Consult https://certbot.eff.org/docs/install.html#problems-with-python-virtual-environment
for possible solutions.
You may also find some support resources at https://certbot.eff.org/support/ .
```

## Certbot's behavior differed from what I expected because:

certbot is not installed

**Comment by SwartzCr** [comment]
It seems like encodings or encodings.idna isn't being imported somewhere, but why I'm not sure.
This is a little outside the realm of my python knowledge but maybe @erikrose or @bmw has an idea why this might happen?

**Comment by SwartzCr** [comment]
It also looks like @fiendish has wrestled with this before (in https://github.com/pyinstaller/pyinstaller/issues/1113) so maybe he would also be able to help?

**Comment by fiendish** [comment]
You can work around it by throwing import encodings.idna at the top of your main py file.

**Comment by fiendish** [comment]
How old is your version of requests? https://github.com/requests/requests/commit/d7227fbb7e07af35f23a0d370ab3b01661af9e40

**Comment by erikrose** [comment]
Old. :-( The latest pip, 9.0.1, which is what certbot-auto pulls in, vendors requests 2.11.1, which doesn't have the patch at https://github.com/requests/requests/commit/d7227fbb7e07af35f23a0d370ab3b01661af9e40.

**Comment by erikrose** [comment]
It looks like the first invocation of pip—the one done by pipstrap, which uses the pip already on your box to bootstrap its way to pip 9.0.1—is going fine. It's the second invocation—the one by certbot-auto directly—that's failing. Unfortuantely, because we run pip in a new process, we have nowhere convenient to put an `import encodings.idna` so that it gets into pip's interpreter's `sys.path`. I see 2 options:

1. Get a new pip released that vendors a newer `requests`.
2. Do something awful. Specifically, make our own Python wrapper for pip which first imports `encodings.idna`.

**Comment by erikrose** [comment]
Looks like pip [has already been updated](https://github.com/pypa/pip/commit/b005a9bb84c1903bf438d57c33938e67e8bb2b29); there just hasn't been a release since the end of 2016. So I vote that, the next time they release, I update pipstrap, roll it in here, and that'll solve this.

**Comment by SwartzCr** [comment]
That sounds good, is there anything this user can do to get around this error?

**Comment by fiendish** [comment]
Well...given that Ubuntu 12.04 has been past end of life for something like 9 months and is therefore no longer receiving security updates...conscience dictates that the user should upgrade their OS instead of looking for support on a system install that smells bad from age. =p

**Comment by erikrose** [comment]
@SwartzCr They could install pip 9.0.1 locally and edit it to include `import encodings.idna` in pip/_vendor/requests/models.py.

**Comment by fdellwing** [comment]
It still gets security updates via the ESM mechanism. But you are correct, the system will be replaced in first quartal 2018, but I just hoped it will be an easy fix.

**Comment by fiendish** [comment]
@erikrose Given that pip only has one code branch on github, and that they're currently sitting on a pile of commits since last release that is now many times larger than the distance between previous releases, it seems like they might not release again for an arbitrarily long time until they figure out how to automate their release process, which is an issue in their tracker dating back to the beginning of 2015. :\

@fdellwing Ah. I forgot about Ubuntu ESM. Can you try what @erikrose is suggesting vis-à-vis modifying models.py in local pip?

**Comment by erikrose** [comment]
I've asked my secret sources when a new pip release might happen. If I hear, you'll hear!

**Comment by fdellwing** [comment]
Ok I got some time for this problem.

First, it was not possible to upgrade pip to 9.0.1 on Ubuntu 12, but the following command did the job `sudo pip install --index-url=https://pypi.python.org/simple/ --upgrade pip`

After that, the `certbot-auto` output does not change, but the problem here is: I cannot edit the models.py file, because it gets downloaded to `/opt/eff.org/certbot` and deleted after it fails.

I than tried updating my `virtualenv` on my local python, but there I got a new problem.

```
[16:43 root@drupal-cms ~] > sudo pip install --upgrade virtualenv
Traceback (most recent call last):
  File "/usr/local/bin/pip", line 9, in <module>
    load_entry_point('pip==9.0.1', 'console_scripts', 'pip')()
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 337, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2279, in load_entry_point
    return ep.load()
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 1989, in load
    entry = __import__(self.module_name, globals(),globals(), ['__name__'])
  File "/usr/local/lib/python2.7/dist-packages/pip/__init__.py", line 21, in <module>
    from pip._vendor.requests.packages.urllib3.exceptions import DependencyWarning
  File "/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/__init__.py", line 66, in <module>
    from .models import Request, Response, PreparedRequest
  File "/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/models.py", line 12, in <module>
    import encodings.idna
  File "/usr/lib/python2.7/encodings/idna.py", line 4, in <module>
    from unicodedata import ucd_3_2_0 as unicodedata
ImportError: cannot import name ucd_3_2_0
```

I'm pretty sure, this is some missing library..

```
[16:46 root@drupal-cms ~] > dpkg -l | grep python
ii  libpython2.7                                   2.7.3-0ubuntu3.9                                      Shared Python runtime library (version 2.7)
ii  python                                         2.7.3-0ubuntu2.2                                      interactive high-level object-oriented language (default version)
ii  python-apport                                  2.0.1-0ubuntu17.15                                    apport crash report handling library
ii  python-apt                                     0.8.3ubuntu7.3                                        Python interface to libapt-pkg
ii  python-apt-common                              0.8.3ubuntu7.3                                        Python interface to libapt-pkg (locales)
ii  python-central                                 0.6.17ubuntu2                                         register and build utility for Python packages
ii  python-chardet                                 2.0.1-2build1                                         universal character encoding detector
ii  python-commando                                0.1.2a-3                                              wrapper for argparse to define declaratively
ii  python-configobj                               4.7.2+ds-3build1                                      simple but powerful config file reader and writer for Python
ii  python-crypto                                  2.4.1-1ubuntu0.1                                      cryptographic algorithms and protocols for Python
ii  python-dbus                                    1.0.0-1ubuntu1                                        simple interprocess messaging system (Python interface)
ii  python-dbus-dev                                1.0.0-1ubuntu1                                        main loop integration development files for python-dbus
ii  python-debian                                  0.1.21ubuntu1                                         Python modules to work with Debian-related data formats
ii  python-dev                                     2.7.3-0ubuntu2.2                                      header files and a static library for Python (default)
ii  python-gamin                                   0.1.10-4ubuntu0.1                                     Python binding for the gamin client library
ii  python-gdbm                                    2.7.3-1ubuntu1                                        GNU dbm database support for Python
ii  python-gi                                      3.2.2-1~precise                                       Python 2.x bindings for gobject-introspection libraries
ii  python-gnupginterface                          0.3.2-9.1ubuntu3                                      Python interface to GnuPG (GPG)
ii  python-httplib2                                0.7.2-1ubuntu2.1                                      comprehensive HTTP client library written for Python
ii  python-keyring                                 0.9.2-0ubuntu0.12.04.2                                store and access your passwords safely
ii  python-launchpadlib                            1.9.12-1                                              Launchpad web services client library
ii  python-lazr.restfulclient                      0.12.0-1ubuntu1.2                                     client for lazr.restful-based web services
ii  python-lazr.uri                                1.0.3-1                                               library for parsing, manipulating, and generating URIs
ii  python-minimal                                 2.7.3-0ubuntu2.2                                      minimal subset of the Python language (default version)
ii  python-newt                                    0.52.11-2ubuntu10                                     A NEWT module for Python
ii  python-oauth                                   1.0.1-3build1                                         Python library implementing of the OAuth protocol
ii  python-openssl                                 0.12-1ubuntu2.1                                       Python wrapper around the OpenSSL library
ii  python-pam                                     0.4.2-12.2ubuntu4                                     A Python interface to the PAM library
ii  python-pip                                     1.0-1build1                                           alternative Python package installer
ii  python-pkg-resources                           0.6.24-1ubuntu1                                       Package Discovery and Resource Access using pkg_resources
ii  python-problem-report                          2.0.1-0ubuntu17.15                                    Python library to handle problem reports
ii  python-pycurl                                  7.19.0-4ubuntu3                                       Python bindings to libcurl
ii  python-serial                                  2.5-2.1build1                                         pyserial - module encapsulating access for the serial port
ii  python-setuptools                              0.6.24-1ubuntu1                                       Python Distutils Enhancements (setuptools compatibility)
ii  python-simplejson                              2.3.2-1                                               simple, fast, extensible JSON encoder/decoder for Python
ii  python-software-properties                     0.82.7.7                                              manage the repositories that you install software from
ii  python-support                                 1.0.14ubuntu2                                         automated rebuilding support for Python modules
ii  python-twisted-bin                             11.1.0-1ubuntu2                                       Event-based framework for internet applications
ii  python-twisted-core                            11.1.0-1ubuntu2                                       Event-based framework for internet applications
ii  python-virtualenv                              1.7.1.2-1                                             Python virtual environment creator
ii  python-wadllib                                 1.3.0-2                                               Python library for navigating WADL files
ii  python-xapian                                  1.2.8-1                                               Xapian search engine interface for Python
ii  python-zope.interface                          3.6.1-1ubuntu3                                        Interfaces for Python
ii  python2.7                                      2.7.3-0ubuntu3.9                                      Interactive high-level object-oriented language (version 2.7)
ii  python2.7-dev                                  2.7.3-0ubuntu3.9                                      Header files and a static library for Python (v2.7)
ii  python2.7-minimal                              2.7.3-0ubuntu3.9                                      Minimal subset of the Python language (version 2.7)
```

**Comment by erikrose** [comment]
If the version of `pip` on the PATH is 9.0.1 or better, pipstrap should do nothing, and your already-installed version of pip should run. And, in fact, that is what we see from the pathnames in your traceback. So you've got that part right, and your models.py patch is working. But now it looks like you've hit a second problem.

**Comment by schoen** [comment]
@erikrose, could you please confirm whether this has subsequently been fixed?

**Comment by erikrose** [comment]
I don't think it has, unless by that version of Ubuntu passing into disuse. pipstrap still installs pip 9.0.1 (though many newer versions have since been released), and, in fact, it now *always* installs 9.0.1, even if a newer version is already present. (We ran into some bugs with newer pips and decided it was safer to use a known-good version.) We ought to update pipstrap's version of pip one of these days.
